File: core/envelope-sim/src/test/java/fr/sncf/osrd/envelope_sim/MaxSpeedEnvelopeTest.java
package fr.sncf.osrd.envelope_sim;

import static fr.sncf.osrd.envelope.EnvelopeShape.CONSTANT;
import static fr.sncf.osrd.envelope.EnvelopeShape.DECREASING;
import static fr.sncf.osrd.envelope_sim.SimpleContextBuilder.TIME_STEP;
import static fr.sncf.osrd.envelope_sim.SimpleContextBuilder.makeSimpleContext;
import static fr.sncf.osrd.envelope_sim.TestMRSPBuilder.makeComplexMRSP;
import static fr.sncf.osrd.envelope_sim.TestMRSPBuilder.makeSimpleMRSP;

import fr.sncf.osrd.envelope.EnvelopeShape;
import fr.sncf.osrd.envelope.EnvelopeTransitions;
import fr.sncf.osrd.envelope_sim.pipelines.MaxSpeedEnvelope;
import org.junit.jupiter.api.Test;


public class MaxSpeedEnvelopeTest {
    @Test
    public void testFlat() {
        var testContext = makeSimpleContext(100000, 0);
        var stops = new double[] { 8500 };

        var flatMRSP = makeSimpleMRSP(testContext, 44.4);
        var context = makeSimpleContext(100000, 0);
        var maxSpeedEnvelope = MaxSpeedEnvelope.from(context, stops, flatMRSP);
        EnvelopeShape.check(maxSpeedEnvelope, CONSTANT, DECREASING, CONSTANT);
        var delta = 2 * maxSpeedEnvelope.getMaxSpeed() * TIME_STEP;
        // don't modify these values, they have been calculated with a 0.1s timestep so they can be considered as
        // reference, the delta is supposed to absorb the difference for higher timesteps
        EnvelopeTransitions.checkPositions(maxSpeedEnvelope, delta, 6529, 8500);
        EnvelopeTransitions.checkContinuity(maxSpeedEnvelope, true, false);
    }

    @Test
    public void testSteep() {
        var testContext = makeSimpleContext(10000, 20);
        var stops = new double[] { 8500 };

        var flatMRSP = makeSimpleMRSP(testContext, 44.4);
        var context = makeSimpleContext(10000, 20);
        var maxSpeedEnvelope = MaxSpeedEnvelope.from(context, stops, flatMRSP);
        EnvelopeShape.check(maxSpeedEnvelope, CONSTANT, DECREASING, CONSTANT);
        var delta = 2 * maxSpeedEnvelope.getMaxSpeed() * TIME_STEP;
        // don't modify these values, they have been calculated with a 0.1s timestep so they can be considered as
        // reference, the delta is supposed to absorb the difference for higher timesteps
        EnvelopeTransitions.checkPositions(maxSpeedEnvelope, delta, 6529, 8500);
        EnvelopeTransitions.checkContinuity(maxSpeedEnvelope, true, false);
    }

    @Test
    public void testInitialStop() {
        var testContext = makeSimpleContext(10000, 0);
        var stops = new double[]{0};

        var flatMRSP = makeSimpleMRSP(testContext, 44.4);
        var context = makeSimpleContext(10000, 0);
        var maxSpeedEnvelope = MaxSpeedEnvelope.from(context, stops, flatMRSP);
        EnvelopeShape.check(maxSpeedEnvelope, CONSTANT);
    }

    @Test
    public void testFlatNonConstDec() {
        var testRollingStock = SimpleRollingStock.MAX_DEC_TRAIN;
        var effortCurveMap = SimpleRollingStock.LINEAR_EFFORT_CURVE_MAP;
        var testPath = new FlatPath(10000, 0);
        var testContext = new EnvelopeSimContext(testRollingStock, testPath, TIME_STEP, effortCurveMap);
        var stops = new double[] { 8500 };

        var flatMRSP = makeSimpleMRSP(testContext, 44.4);
        var context = new EnvelopeSimContext(testRollingStock, testPath, TIME_STEP, effortCurveMap);
        var maxSpeedEnvelope = MaxSpeedEnvelope.from(context, stops, flatMRSP);
        EnvelopeShape.check(maxSpeedEnvelope, CONSTANT, DECREASING, CONSTANT);
        var delta = 2 * maxSpeedEnvelope.getMaxSpeed() * TIME_STEP;
        // don't modify these values, they have been calculated with a 0.1s timestep so they can be considered as
        // reference, the delta is supposed to absorb the difference for higher timesteps
        EnvelopeTransitions.checkPositions(maxSpeedEnvelope, delta, 7493, 8500);
        EnvelopeTransitions.checkContinuity(maxSpeedEnvelope, true, false);
    }

    @Test
    public void testWithComplexMRSP() {
        var length = 100000;
        var testContext = makeSimpleContext(length, 0);
        var stops = new double[] {50000, length};

        var mrsp = makeComplexMRSP(testContext);
        var context = makeSimpleContext(length, 0);
        var maxSpeedEnvelope = MaxSpeedEnvelope.from(context, stops, mrsp);
        EnvelopeShape.check(maxSpeedEnvelope, CONSTANT, CONSTANT, DECREASING, CONSTANT,
                DECREASING, CONSTANT, CONSTANT, CONSTANT, DECREASING, CONSTANT, DECREASING);
        EnvelopeTransitions.checkContinuity(maxSpeedEnvelope,
                false, true, true, true, false, false, false, true, true, true);
    }
}

File: core/envelope-sim/src/test/kotlin/fr/sncf/osrd/envelope_sim/MaxSpeedEnvelopeTest.kt
package fr.sncf.osrd.envelope_sim

import fr.sncf.osrd.envelope.EnvelopeShape
import fr.sncf.osrd.envelope.EnvelopeTransitions
import fr.sncf.osrd.envelope_sim.pipelines.SimStop
import fr.sncf.osrd.envelope_sim.pipelines.maxSpeedEnvelopeFrom
import fr.sncf.osrd.railjson.schema.schedule.RJSTrainStop.RJSReceptionSignal
import fr.sncf.osrd.utils.units.Offset
import fr.sncf.osrd.utils.units.meters
import org.junit.jupiter.api.Test

class MaxSpeedEnvelopeTest {
    @Test
    fun testFlat() {
        val testContext = SimpleContextBuilder.makeSimpleContext(100000.0, 0.0)
        val stops = listOf(SimStop(Offset(8500.0.meters), RJSReceptionSignal.SHORT_SLIP_STOP))

        val flatMRSP = TestMRSPBuilder.makeSimpleMRSP(testContext, 44.4)
        val context = SimpleContextBuilder.makeSimpleContext(100000.0, 0.0)
        val maxSpeedEnvelope = maxSpeedEnvelopeFrom(context, stops, flatMRSP)
        EnvelopeShape.check(
            maxSpeedEnvelope,
            EnvelopeShape.CONSTANT,
            EnvelopeShape.DECREASING,
            EnvelopeShape.CONSTANT,
        )
        val delta = 2 * maxSpeedEnvelope.maxSpeed * SimpleContextBuilder.TIME_STEP
        // don't modify these values, they have been calculated with a 0.1s timestep so they can be
        // considered as
        // reference, the delta is supposed to absorb the difference for higher timesteps
        EnvelopeTransitions.checkPositions(maxSpeedEnvelope, delta, 6529.0, 8500.0)
        EnvelopeTransitions.checkContinuity(maxSpeedEnvelope, true, false)
    }

    @Test
    fun testSteep() {
        val testContext = SimpleContextBuilder.makeSimpleContext(10000.0, 20.0)
        val stops = listOf(SimStop(Offset(8500.meters), RJSReceptionSignal.STOP))

        val flatMRSP = TestMRSPBuilder.makeSimpleMRSP(testContext, 44.4)
        val context = SimpleContextBuilder.makeSimpleContext(10000.0, 20.0)
        val maxSpeedEnvelope = maxSpeedEnvelopeFrom(context, stops, flatMRSP)
        EnvelopeShape.check(
            maxSpeedEnvelope,
            EnvelopeShape.CONSTANT,
            EnvelopeShape.DECREASING,
            EnvelopeShape.CONSTANT,
        )
        val delta = 2 * maxSpeedEnvelope.maxSpeed * SimpleContextBuilder.TIME_STEP
        // don't modify these values, they have been calculated with a 0.1s timestep so they can be
        // considered as
        // reference, the delta is supposed to absorb the difference for higher timesteps
        EnvelopeTransitions.checkPositions(maxSpeedEnvelope, delta, 6529.0, 8500.0)
        EnvelopeTransitions.checkContinuity(maxSpeedEnvelope, true, false)
    }

    @Test
    fun testInitialStop() {
        val testContext = SimpleContextBuilder.makeSimpleContext(10000.0, 0.0)
        val stops = listOf(SimStop(Offset(0.meters), RJSReceptionSignal.OPEN))

        val flatMRSP = TestMRSPBuilder.makeSimpleMRSP(testContext, 44.4)
        val context = SimpleContextBuilder.makeSimpleContext(10000.0, 0.0)
        val maxSpeedEnvelope = maxSpeedEnvelopeFrom(context, stops, flatMRSP)
        EnvelopeShape.check(maxSpeedEnvelope, EnvelopeShape.CONSTANT)
    }

    @Test
    fun testFlatNonConstDec() {
        val testRollingStock = SimpleRollingStock.MAX_DEC_TRAIN
        val effortCurveMap = SimpleRollingStock.LINEAR_EFFORT_CURVE_MAP
        val testPath = FlatPath(10000.0, 0.0)
        val testContext =
            EnvelopeSimContext(
                testRollingStock,
                testPath,
                SimpleContextBuilder.TIME_STEP,
                effortCurveMap,
            )
        val stops = listOf(SimStop(Offset(8500.meters), RJSReceptionSignal.SHORT_SLIP_STOP))

        val flatMRSP = TestMRSPBuilder.makeSimpleMRSP(testContext, 44.4)
        val context =
            EnvelopeSimContext(
                testRollingStock,
                testPath,
                SimpleContextBuilder.TIME_STEP,
                effortCurveMap,
            )
        val maxSpeedEnvelope = maxSpeedEnvelopeFrom(context, stops, flatMRSP)
        EnvelopeShape.check(
            maxSpeedEnvelope,
            EnvelopeShape.CONSTANT,
            EnvelopeShape.DECREASING,
            EnvelopeShape.CONSTANT,
        )
        val delta = 2 * maxSpeedEnvelope.maxSpeed * SimpleContextBuilder.TIME_STEP
        // don't modify these values, they have been calculated with a 0.1s timestep so they can be
        // considered as
        // reference, the delta is supposed to absorb the difference for higher timesteps
        EnvelopeTransitions.checkPositions(maxSpeedEnvelope, delta, 7493.0, 8500.0)
        EnvelopeTransitions.checkContinuity(maxSpeedEnvelope, true, false)
    }

    @Test
    fun testWithComplexMRSP() {
        val length = 100000
        val testContext = SimpleContextBuilder.makeSimpleContext(length.toDouble(), 0.0)
        val stops =
            listOf(
                SimStop(Offset(50000.meters), RJSReceptionSignal.SHORT_SLIP_STOP),
                SimStop(Offset(length.meters), RJSReceptionSignal.SHORT_SLIP_STOP),
            )

        val mrsp = TestMRSPBuilder.makeComplexMRSP(testContext)
        val context = SimpleContextBuilder.makeSimpleContext(length.toDouble(), 0.0)
        val maxSpeedEnvelope = maxSpeedEnvelopeFrom(context, stops, mrsp)
        EnvelopeShape.check(
            maxSpeedEnvelope,
            EnvelopeShape.CONSTANT,
            EnvelopeShape.CONSTANT,
            EnvelopeShape.DECREASING,
            EnvelopeShape.CONSTANT,
            EnvelopeShape.DECREASING,
            EnvelopeShape.CONSTANT,
            EnvelopeShape.CONSTANT,
            EnvelopeShape.CONSTANT,
            EnvelopeShape.DECREASING,
            EnvelopeShape.CONSTANT,
            EnvelopeShape.DECREASING,
        )
        EnvelopeTransitions.checkContinuity(
            maxSpeedEnvelope,
            false,
            true,
            true,
            true,
            false,
            false,
            false,
            true,
            true,
            true,
        )
    }
}
